From ea84ec9583667b557895789cb1fdf33219ad900d Mon Sep 17 00:00:00 2001
From: Liu Ying <victor.liu@nxp.com>
Date: Mon, 6 Nov 2017 10:35:46 +0800
Subject: [PATCH 2736/5242] MLK-16772 gpu: imx: dpu: cf: Add
 safety_stream_cf_color module parameter support

commit  d6b308c9c6efb68bc976f4a1cea4ba05d40ade36 from
https://source.codeaurora.org/external/imx/linux-imx.git

This patch adds safety_stream_cf_color module parameter support so that
users may set the default color generated by the constframe units(4 and 5)
of safety streams.

Signed-off-by: Liu Ying <victor.liu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/imx/dpu/dpu-crtc.c   |    1 -
 drivers/gpu/imx/dpu/dpu-constframe.c |   16 +++++++++++++++-
 2 files changed, 15 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/imx/dpu/dpu-crtc.c b/drivers/gpu/drm/imx/dpu/dpu-crtc.c
index 506e12e..3710333 100644
--- a/drivers/gpu/drm/imx/dpu/dpu-crtc.c
+++ b/drivers/gpu/drm/imx/dpu/dpu-crtc.c
@@ -467,7 +467,6 @@ static void dpu_crtc_mode_set_nofb(struct drm_crtc *crtc)
 	constframe_framedimensions(dpu_crtc->cf,
 					mode->crtc_hdisplay,
 					mode->crtc_vdisplay);
-	constframe_constantcolor(dpu_crtc->cf, 0, 0, 0, 0);
 
 	ed_src = dpu_crtc->stream_id ? ED_SRC_CONSTFRAME5 : ED_SRC_CONSTFRAME4;
 	extdst_pixengcfg_src_sel(dpu_crtc->ed, ed_src);
diff --git a/drivers/gpu/imx/dpu/dpu-constframe.c b/drivers/gpu/imx/dpu/dpu-constframe.c
index 1290545..d014830 100644
--- a/drivers/gpu/imx/dpu/dpu-constframe.c
+++ b/drivers/gpu/imx/dpu/dpu-constframe.c
@@ -21,6 +21,11 @@
 #include <video/dpu.h>
 #include "dpu-prv.h"
 
+static unsigned int safety_stream_cf_color = 0x0;
+module_param(safety_stream_cf_color, uint, 0444);
+MODULE_PARM_DESC(safety_stream_cf_color,
+"Safety stream constframe color in hex(0xRRGGBBAA) [default=0x00000000]");
+
 #define FRAMEDIMENSIONS		0xC
 #define WIDTH(w)		(((w) - 1) & 0x3FFF)
 #define HEIGHT(h)		((((h) - 1) & 0x3FFF) << 16)
@@ -168,6 +173,7 @@ void dpu_cf_put(struct dpu_constframe *cf)
 
 void _dpu_cf_init(struct dpu_soc *dpu, unsigned int id)
 {
+	struct dpu_constframe *cf;
 	int i;
 
 	for (i = 0; i < ARRAY_SIZE(cf_ids); i++)
@@ -177,7 +183,15 @@ void _dpu_cf_init(struct dpu_soc *dpu, unsigned int id)
 	if (WARN_ON(i == ARRAY_SIZE(cf_ids)))
 		return;
 
-	constframe_shden(dpu->cf_priv[i], true);
+	cf = dpu->cf_priv[i];
+
+	constframe_shden(cf, true);
+
+	if (id == 4 || id == 5) {
+		mutex_lock(&cf->mutex);
+		dpu_cf_write(cf, safety_stream_cf_color, CONSTANTCOLOR);
+		mutex_unlock(&cf->mutex);
+	}
 }
 
 int dpu_cf_init(struct dpu_soc *dpu, unsigned int id,
-- 
1.7.9.5

